Plugins/Public/playercntl_plugin/MiscCmds.cpp

```diff
@@ -217,23 +226,99 @@ namespace MiscCmds
-		wscMsg = ReplaceStr(wscMsg, L"%max", stows(itos(max)));
-		PrintLocalUserCmdText(iFromClientID, wscMsg, set_iLocalChatRange);
-		return true;
-	}
+	/*
+	Roll dice for everyone within 6km of a vessel. Supports 1d20 formatting.
+	*/
+	bool MiscCmds::UserCmd_Dice(uint iFromClientID, const wstring &wscCmd, const wstring &wscParam, const wchar_t *usage)
+	{

+		boost::wregex expr(L"(\\d{1,2})[Dd](\\d{1,3})(([+\\-*])?(\\d{1,5}))?");
+		boost::wsmatch sm;
+
+		// If the regex finds a match denoting the correct roll format, run the randomized numbers
+		if (boost::regex_match(wscParam, sm, expr))
+		{
+
+			// Smatch index [1] represents the roll count
+			int rollCount = _wtoi(sm[1].str().c_str());
+
+			// Smatch index [2] represents the dice count
+			int diceCount = _wtoi(sm[2].str().c_str());
+
+			// Smatch index [3] represents any modifier numeric value. This is set ONLY if we are using a mod-operation
+			int modifierValue;
+
+			diceOperation operation;
+			if (sm[3].str().find(L"+") == 0)
+			{
+				operation = diceOperation::ADD;
+				modifierValue = _wtoi(sm[5].str().c_str());
+			}
+			else if (sm[3].str().find(L"-") == 0)
+			{
+				operation = diceOperation::SUBTRACT;
+				modifierValue = _wtoi(sm[5].str().c_str());
+			}
+			else
+			{
+				operation = diceOperation::NONE;
+			}
+
+			string diceResultSteps = "";
+			uint number = 0;
+
+			for (int i = 0; i < rollCount; i++)
+			{
+				int randValue = (rand() % diceCount) + 1;
+
+				// If we have a modifier, apply it
+				if (operation == diceOperation::ADD)
+				{
+					number += (randValue + modifierValue);
+					diceResultSteps.append("(").append(itos(randValue)).append(" + ").append(itos(modifierValue).append(")"));
+				}
+				else if (operation == diceOperation::SUBTRACT)
+				{
+					number += (randValue - modifierValue);
+					diceResultSteps.append("(").append(itos(randValue)).append(" - ").append(itos(modifierValue).append(")"));
+				}
+				else
+				{
+					number += randValue;
+					diceResultSteps.append("(").append(itos(randValue)).append(")");
+				}
+
+				// Are we not on the last value? Keep the string pretty by adding another +
+				if (i < rollCount - 1)
+				{
+					diceResultSteps.append(" + ");
+				}
+			}
+
+			wstring wscCharname = (const wchar_t*)Players.GetActiveCharacterName(iFromClientID);
+
+			// Print the results
+			wstring diceAlert = L"%player rolled %value with the formula %formula";
+			diceAlert = ReplaceStr(diceAlert, L"%player", wscCharname);
+			diceAlert = ReplaceStr(diceAlert, L"%value", stows(itos(number)));
+			diceAlert = ReplaceStr(diceAlert, L"%formula", sm[0].str().c_str());
+
+			PrintLocalUserCmdText(iFromClientID, diceAlert, set_iLocalChatRange);
+
+			// Only print the steps taken if less than 10 dice was rolled.
+			if (rollCount < 10)
+			{
+				PrintLocalUserCmdText(iFromClientID, stows(diceResultSteps), set_iLocalChatRange);
+			}
+
+		}
+		else
+		{
+			PrintUserCmdText(iFromClientID, L"Usage: /roll 1d20");

```

Comment:
Some comments about what each means would be nice

Reply:
Done.

```
PrintUserCmdText(iFromClientID, L"Usage: (NumDice) d (DiceSides) [+-] (Modifier)");
PrintUserCmdText(iFromClientID, L"Examples: /roll 1d20 -- Roll 1, 20 sided die");
PrintUserCmdText(iFromClientID, L"          /roll 1d8+4 -- Roll 1, 8 sided die and add 8");
PrintUserCmdText(iFromClientID, L"          /roll 4d20+2 -- Roll 4, 20 sided dice, adding 2 to each die rolled");
```

Reply:
Nitpick: 'Roll 1, 8 sided dice, adding 8 to each die rolled' ?